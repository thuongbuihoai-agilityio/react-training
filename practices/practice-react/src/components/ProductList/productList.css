/* style for product list */
.product-list-normal {
  width: 384px;
  margin: 100px 70px 188px 0;
  background: var(--light);
  box-shadow: 0px 10px 40px rgba(0, 0, 0, 0.1);
}

.product-list-medium {
  width: 368px;
  border-radius: 20px;
  margin: 50px 85px 164px 0;
  background: var(--light);
}

.product-list-medium-box {
  width: 368px;
  background: var(--light);
  border-radius: 10px;
  margin: 0 80px 159px 0;
  box-shadow: 0px 16px 70px var(--box-first);
}

/* style for card in product list */
.cart-link {
  text-decoration: none;
  color: var(--text-primary);
}

.card-discount {
  padding: 59px 0 60px 44px;
}

.card-quantity {
  padding: 57px 0 15px 40px;
}

.card-unit {
  margin: 9px 0 14px;
  color: var(--bg-secondary);
}

.card-image {
  width: 368px;
  height: 288px;
}

.card-item-offers,
.card-item-selling,
.card-item-popular {
  display: flex;
  flex-wrap: wrap;
}

.card-item-offers {
  margin: 33px 0 0 90px;
}

.card-item-selling,
.card-item-popular {
  padding-left: 100px;
}

.card-counter {
  display: flex;
  justify-content: space-between;
  margin: 0 35px 18px;
}

.card-offers-list-title {
  margin: 33px 0 0 130px
}

.card-popular-title {
  margin: 205px 0 44px 103px
}

.card-product-selling-title {
  padding: 84px 0 49px 125px;
}

.card-product-selling-list {
  background: var(--bg-light);
}