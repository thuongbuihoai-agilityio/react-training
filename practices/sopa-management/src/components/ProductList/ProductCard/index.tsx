import { memo } from 'react';

// Components
import Card from '@components/Card';

// Interfaces
import { Product } from '@interfaces/product';

// Styles
import './productCard.css';

interface ProductCardProps {
  key?: number;
  data: Product[];
}

const ProductCard = ({
  data,
}: ProductCardProps) => (
  <div className='product-card' data-testid='product-card'>
    {data?.map((product) => (
      <Card
        key={product.id}
        href={`products/${product.id}`}
        src={product.image.url}
        name={product.name}
        color={product.color}
        price={product.price}
        alt={product.image.alt}
      />
    ))}
  </div>
);

export default memo(ProductCard);
