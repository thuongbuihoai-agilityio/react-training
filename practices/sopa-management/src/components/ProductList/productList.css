.product {
  padding: var(--px-medium) var(--py-medium);
}

.product-btn {
  text-align: center;
}

.product-item-btn {
  font-family: 'var(--font-primary)';
  font-size: var(--text-extra-sm);
}
