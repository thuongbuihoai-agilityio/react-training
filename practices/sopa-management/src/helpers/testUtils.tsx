import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import {
  QueryClient,
  QueryClientProvider
} from 'react-query';
import {
  BrowserRouter,
  Router
} from 'react-router-dom';

export const renderRouterTest = (component: JSX.Element) => {
  const history = createMemoryHistory();
  return render(
    <Router location={history.location} navigator={history}>
      {component}
    </Router>
  );
};

export const renderWithRouterAndQuery = (children: JSX.Element) => {
  const queryClient = new QueryClient();

  return render(
    <BrowserRouter>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </BrowserRouter>
  );
};
