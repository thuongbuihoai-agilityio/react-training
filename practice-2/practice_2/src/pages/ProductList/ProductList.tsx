import React, { useEffect, useState } from "react";
import axios from "axios";
import { cardImage } from "../../constants/card";
import Price from "../../components/Price/Price";
import Products from "../../components/ProductItem/Product";
import Title from "../../components/common/Title/Title";
import "./productList.css"
import { useLocation } from "react-router-dom";

export default function ProductList() {
  const [products, setProducts] = useState([]);

  const location = useLocation()
  const category = location.state?location.state:""
  const image = location.state?location.state:""

  useEffect(() => {
    async function fetchMyAPI() {
      const result = await axios
      .get("products")
      .then(function (response) {
        console.log(response);
        setProducts(response.data);
      })
      .catch(function (error) {
        console.log(error);
      });
    }
    fetchMyAPI()
  }, [category])

  return (
    products.map(product =>
      <div className="product__list" key={product.id}>
        <div className="product__item">
          <Products product={product} />
          <div className="product__content">
            <Title className={""} value={product.name} />
            <Price value={product.price} />
          </div>
        </div>
      </div>
    )
  );
}