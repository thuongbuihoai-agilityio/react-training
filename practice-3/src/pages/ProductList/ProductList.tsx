import React, { useCallback, useState } from "react";
import ModalCreate from "@/components/Modal/ModalCreate/ModalCreate";
import Button from "@/components/common/Button/Button";
import useProducts from "@/hooks/useProducts";
import { Product } from "@/types/product";
import ProductItem from "../ProductItem/ProductItem";
import "./productList.css";

const ProductList: React.FC = () => {
  const { products, deleteProduct } = useProducts();
  const [openModalUpdate, setOpenModalUpdate] = useState(false);

  const handleOpenModalUpdate = useCallback(() => {
    setOpenModalUpdate(true);
  }, []);

  return (
    <>
      <div data-testid="product-list-page" className="product__list">
        <Button onClick={handleOpenModalUpdate} className="btn btn__add" text="Add new product" />
        <div className="product__info">
          {products?.map((product: Product) =>
            <div className="product__item" key={product.id}>
              <ProductItem deleteProduct={deleteProduct} product={product} />
            </div>
          )}
        </div>
        {openModalUpdate && <ModalCreate hideModalUpdate={setOpenModalUpdate} />}
      </div>
    </>
  );
}

export default ProductList;
