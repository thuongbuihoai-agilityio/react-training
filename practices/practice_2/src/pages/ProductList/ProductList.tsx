import React from "react";
import Title from "../components/Title/Title";
import Price from "../components/Price/Price";
import ProductItem from "@/components/ProductItem/ProductItem";
import { Product } from "@/types/product";
import "./productList.css";
import useProduct from "@/hooks/useProduct";

const ProductList: React.FC = () => {
  const { products } = useProduct();
  return (
    <div className="product__list">
      {products.map((product: Product, index: number) =>
        <div className="product__item" key={index}>
          <ProductItem product={product} />
          <div className="product__content">
            <Title text={product.name} />
            <Price value={product.price.toString()} />
          </div>
        </div>
      )}
    </div>
  );
}

export default ProductList;
