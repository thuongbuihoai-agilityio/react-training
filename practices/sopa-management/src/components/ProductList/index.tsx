// Hooks
import { useInfiniteProducts } from '@hooks/useQuery';

// Components
import {
  Button,
  ButtonType,
  Loading
} from '@components/common';
import ProductCard from './ProductCard';

// Styles
import './productList.css';

const ProductList = () => {
  const { data, fetchNextPage, hasNextPage, isLoading, isFetchingNextPage } =
    useInfiniteProducts();

    return (
    <div data-testid='product' className='product'>
      {isLoading ? (
        <Loading />
      ) : (
        <>
          <div data-testid='product-list' className='product-list'>
            {data?.map((page, index) =>
                <ProductCard
                  key={index}
                  data={page}
                />
            )}
          </div>
          <div className='product-btn'>
            {hasNextPage && (
              <Button
                ariaLabel='Show more'
                className='product-item-btn'
                children={isFetchingNextPage ? 'Loading more...' : 'Show More'}
                type={ButtonType.primary}
                onClick={() => fetchNextPage()}
                disable={isFetchingNextPage}
              />
            )}
          </div>
        </>
      )}
    </div>
  );
};

export default ProductList;
