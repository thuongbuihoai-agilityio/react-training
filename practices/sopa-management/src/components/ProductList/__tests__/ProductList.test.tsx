// Libs
import '@testing-library/jest-dom';
import {
  waitFor,
  screen
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// Components
import ProductList from '@components/ProductList';

// Helpers
import { renderWithRouterAndQuery } from '@helpers/testUtils';

const fetchNextPage = jest.fn();
jest.mock('@hooks/useQuery', () => ({
  useInfiniteProducts: () => ({
    fetchNextPage,
    hasNextPage: true,
    isLoading: false,
    isFetchingNextPage: false,
  }),
}));

describe('ProductList component', () => {
  test('should render ProductList component', () => {
    const { getByTestId } = renderWithRouterAndQuery(<ProductList />);
    expect(getByTestId('product')).toBeInTheDocument();
  });

  test('renders loading state initially', async () => {
    const { getByTestId } = renderWithRouterAndQuery(<ProductList />);

    expect(getByTestId('product')).toBeInTheDocument();
    await waitFor(() => screen.getByTestId('product-list'));
    expect(screen.queryByText('Loading...')).not.toBeInTheDocument();
    expect(getByTestId('product-list')).toBeInTheDocument();
  });

  test('calls fetchNextPage on button click', async () => {
    renderWithRouterAndQuery(<ProductList />);

    const showMoreButton =  screen.getByText('Show More');
    await userEvent.click(showMoreButton);

    await waitFor(() => expect(fetchNextPage).toHaveBeenCalled());
  });

  test('matches snapshot', () => {
    const { container } = renderWithRouterAndQuery(<ProductList />);
    expect(container).toMatchSnapshot();
  });
});
