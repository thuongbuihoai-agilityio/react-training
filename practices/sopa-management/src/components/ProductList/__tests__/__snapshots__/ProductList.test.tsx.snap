// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList component matches snapshot 1`] = `
<div>
  <div
    class="product"
    data-testid="product"
  >
    <div
      class="product-list"
      data-testid="product-list"
    />
    <div
      class="product-btn"
    >
      <button
        aria-label="Show more"
        class="product-item-btn btn btn-primary"
        data-testid="button"
      >
        Show More
      </button>
    </div>
  </div>
</div>
`;
